### ISSUE TITLE
Export ONNX with dynamic batchsize and NMS

### ISSUE THREAD

**Issue by Egorundel** [ultralytics/ultralytics#23339]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Hi everyone! 

I'm using inference on TensorRT models. But to do this, I need to create an ONNX model with dynamic BatchSize and NMS outputs, so my inference is written in C++ using NMS output processing. Why doesn't the standard code work with NMS for end-to-end models? Is there a script that can create an ONNX model with NMS from `model.pt`?

### Use case

`export_yolo26.py`:
```python
from ultralytics import YOLO

model = YOLO("./yolo26m.pt")
model.export(format="onnx", imgsz=640, dynamic=True, simplify=True, nms=True, batch=12)
```

Command:
```cmd
python3 export_yolo26.py
```

Output:
```cmd
WARNING ⚠ 'nms=True' is not available for end2end models. Forcing 'nms=False'.
```

### Additional

default YOLO26 (m) model

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Egorundel, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it (a tiny script + exact command + full console output). In your case, please also include:
-Exact `ultralytics` version (`pip show ultralytics`)
-Your OS, Python, and PyTorch versions
-The exact `model.export(...)` arguments you’re using (as you shared) and whether the model is end-to-end (as indicated by the warning)
-The resulting exported artifacts and logs (ONNX export log lines)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

⚙️ Note: This is an automated response to help gather the right details faster. An Ultralytics engineer will also assist soon 🙏

**Comment by Y-T-G** [comment]
YOLO26 doesn't use NMS which is why it doesn't work

**Comment by sujunze** [comment]
After I exported it as engine model, the conf value in the inference result is very different from the conf value in the inference result of pt model. The result of pt inference is 0.35. The engine reasoned that it was 0.18

**Comment by sujunze** [comment]
I also trained the model on yolo26.pt model，We hope that the author can provide a perfect way to export the model that can be accelerated using tensorrt

**Comment by Egorundel** [comment]
> YOLO26 не использует NMS, поэтому он не работает

How do you deal with multiple detections so that they don't overlap?

**Comment by Y-T-G** [comment]
> After I exported it as engine model, the conf value in the inference result is very different from the conf value in the inference result of pt model. The result of pt inference is 0.35. The engine reasoned that it was 0.18

PyTorch model runs with minimum rectangle padding. You can get the same result if you export with `dynamic=True`

**Comment by Y-T-G** [comment]
> > YOLO26 не использует NMS, поэтому он не работает
> 
> How do you deal with multiple detections so that they don't overlap?

Are you getting that with YOLO26?

**Comment by Pro100rus32** [comment]
> Are you getting that with YOLO26?

Yes, it seems there is some kind of problem with the built-in nms, it can produce 2 or more detections with almost 99% overlap with onnx.

**Comment by Y-T-G** [comment]
@Pro100rus32 Are they the same class or different class?

**Comment by Pro100rus32** [comment]
> [@Pro100rus32](https://github.com/Pro100rus32) Are they the same class or different class?

Same class

**Comment by glenn-jocher** [comment]
If you’re seeing near-identical same-class boxes from `yolo26*.onnx`, it’s usually a postprocess/parsing mismatch (e.g., treating the raw output tensor as “final detections” without the same decode/filter logic Ultralytics applies); can you sanity-check by running the exported ONNX through Ultralytics and confirm whether duplicates still appear?

```python
from ultralytics import YOLO

pt = YOLO("yolo26m.pt")
onnx = YOLO("yolo26m.onnx")

pt.predict("image.jpg", imgsz=640, conf=0.25)
onnx.predict("image.jpg", imgsz=640, conf=0.25)
```

If duplicates show up in the ONNX path here too, please paste `yolo checks` output plus a minimal repro image and your export args; if they *don’t*, then the fix is in your C++ postprocess (decode/thresholding), and note that YOLO26 is designed to be NMS-free so the exporter intentionally won’t embed NMS into ONNX for end-to-end models (see the [NMS-free YOLO26 deployment note](https://www.ultralytics.com/blog/why-ultralytics-yolo26-removes-nms-and-how-that-changes-deployment) and the [export docs](https://docs.ultralytics.com/modes/export/)).
